packages/core/src/utils/userop.ts

```diff
@@ -1,32 +1,104 @@
+} from "viem";
+import type { EntryPointVersion } from "../entrypoint/types";
+import { UnexpectedEntryPointVersionError } from "../errors/entrypoint.js";
 import type {
   BigNumberish,
   Multiplier,
   UserOperationFeeOptionsField,
   UserOperationRequest,
+  UserOperationRequest_v7,
   UserOperationStruct,
+  UserOperationStruct_v6,
+  UserOperationStruct_v7,
 } from "../types";
 import { bigIntClamp, bigIntMultiply } from "./bigint.js";
-import { isBigNumberish } from "./index.js";
+import { allEqual, isBigNumberish } from "./index.js";
 
 /**
- * Utility method for asserting a {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * Utility method for asserting a {@link UserOperationStruct} has valid fields for the given entry point version
  *
  * @param request a {@link UserOperationStruct} to validate
- * @returns a type guard that asserts the {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * @param entryPointVersion a {@link EntryPointVersion} entry point version
+ * @returns a type guard that asserts the {@link UserOperationRequest} is valid
  */
-export function isValidRequest(
-  request: UserOperationStruct
-): request is UserOperationRequest {
+export function isValidRequest<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(
+  request: UserOperationStruct<TEntryPointVersion>
+): request is UserOperationRequest<TEntryPointVersion> {
   // These are the only ones marked as optional in the interface above
   return (
-    !!request.callGasLimit &&
-    !!request.maxFeePerGas &&
-    request.maxPriorityFeePerGas != null &&
-    !!request.preVerificationGas &&
-    !!request.verificationGasLimit
+    BigInt(request.callGasLimit || 0n) > 0n &&
+    BigInt(request.maxFeePerGas || 0n) > 0n &&
+    BigInt(request.maxPriorityFeePerGas || 0n) > 0n &&
+    BigInt(request.preVerificationGas || 0n) > 0n &&
+    BigInt(request.verificationGasLimit || 0n) > 0n &&
+    isValidPaymasterAndData(request) &&
+    isValidFactoryAndData(request)
   );
 }
 
+/**
+ * Utility method for asserting a {@link UserOperationRequest} has valid fields for the paymaster data
+ *
+ * @param request a {@link UserOperationRequest} to validate
+ * @param entryPointVersion a {@link EntryPointVersion} entry point version
+ * @returns a type guard that asserts the {@link UserOperationRequest} is a {@link UserOperationRequest}
+ */
+export function isValidPaymasterAndData<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(request: UserOperationStruct<TEntryPointVersion>): boolean {
+  if (!("paymaster" in request)) {
+    const { paymasterAndData } = request as UserOperationStruct_v6;
+    return paymasterAndData != null;
+  }
+
+  const {
+    paymaster,
+    paymasterData,
+    paymasterPostOpGasLimit,
+    paymasterVerificationGasLimit,
+  } = request as UserOperationStruct_v7;
+  // either all exist, or none.
+  return allEqual(
+    isAddress(paymaster),
+    toHex(paymasterData || "0x") !== "0x",
```

Comment:
paymaster fields (paymaster, paymasterData, paymasterValidationGasLimit, paymasterPostOpGasLimit) either all exist, or none.

https://github.com/ethereum/ercs/blob/master/ERCS/erc-4337.md#rpc-methods-eth-namespace.

Reply:
hmm yea but the docs do say this too
> All fields MUST be set as hex values. empty bytes block (e.g. empty initCode) MUST be set to "0x"

reading that, I'd imagine that either all the fields are `0x` or all are `Hex`

Reply:
this is what I referenced: https://github.com/eth-infinitism/account-abstraction/blob/8fae866ea908f517f0399edd27057b2cc275576d/test/UserOp.ts#L28 for packing user operation.

So when packing user operation, if it fails to hold that all paymaster-related fields (4 of them) are not null, paymasterAndData for packed uo is encoded as 0x. But per the user operation request sent to the bundler, they are unset.

cc: @dancoombs @dphilipson can you confirm on this logic?

Reply:
```
**Parameters:**
- UserOperation a full user-operation struct. All fields MUST be set as hex values. empty bytes block (e.g. empty initCode) MUST be set to "0x"
- factory and factoryData - either both exist, or none
- paymaster fields (paymaster, paymasterData, paymasterValidationGasLimit, paymasterPostOpGasLimit) either all exist, or none.
- EntryPoint the entrypoint address the request should be sent through. this MUST be one of the entry points returned by the supportedEntryPoints rpc call.
```

Reply:
When it says "either all exist, or none", that means fields are "unset" or set to "0x" when sending user operation?

Reply:
I do get the user op hash returned by the rundler on staging with unexisting fields unset, and in case of failure, the EIP says it MUST return an error result object, with code: -32602 - invalid UserOperation struct/fields. I don't get this error with the current way of omitting unexisting fields, so I assumed this understanding. cc: @moldy530